### PR TITLE
Add BLS signatures over BLS12-381

### PR DESCRIPTION
**Pull request by armfazh**
Implements the basic version of BLS as in IETF draft.
https://github.com/cfrg/draft-irtf-cfrg-bls-signature

### CODE REVIEW

File: ecc/bls12381/g2.go [hunk 1]
```diff
@@ -187,10 +187,84 @@ func (g *G2) clearCofactor() {
 }
 
 // Double updates g = 2g.
-func (g *G2) Double() { doubleAndLine(g, nil) }
```
Comment (bwesterb) [inline ecc/bls12381/g2.go:190]: Why is this changed?

**Review by thibmeu — COMMENTED** [review]
Tests are passing. Mostly documentation related comments.

I think the code should be better documented. This could be with one big comment at the start with reference to the source algorithm/paper/document, or via multiple inline comments.

File: sign/bls/bls.go [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+type PublicKey[K KeyGroup] struct{ key K }
+
+func (k *PrivateKey[K]) Public() crypto.PublicKey { return k.PublicKey() }
+
+func (k *PrivateKey[K]) PublicKey() *PublicKey[K] {
+	if k.pub == nil {
+		k.pub = new(PublicKey[K])
+		switch (interface{})(k).(type) {
+		case *PrivateKey[G1]:
+			kk := (interface{})(&k.pub.key).(*G1)
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go:52]: it obey to conversions between interfaces to specific structs. The pattern of repeating a letter hints about the origin of the variable.

File: sign/bls/bls.go [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+		default:
+			panic(ErrInvalid)
+		}
+	}
+
+	return k.pub
+}
+
+func (k *PrivateKey[K]) Equal(x crypto.PrivateKey) bool {
+	xx, ok := x.(*PrivateKey[K])
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go:66]: correct, better to fail-fast if `not ok`.

File: sign/bls/bls.go [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+}
+
+var (
+	ErrInvalid   = errors.New("bls: invalid BLS instance")
+	ErrKeyGen    = errors.New("bls: too many unsuccessful key generation tries")
+	ErrShortIKM  = errors.New("bls: IKM material shorter than 32 bytes")
+	ErrAggregate = errors.New("bls: error while aggregating signatures")
+)
+
+const (
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go:334]: no need to be public, as only used internally.

File: ecc/bls12381/g2.go (hunk 1 above)
Comment (armfazh) [reply ecc/bls12381/g2.go]: this is a stale commit, not needed. Fixed.

**Review by thibmeu — APPROVED** [review]
Good codewise for me.

File: sign/bls/bls.go [hunk 5]
```diff
@@ -0,0 +1,414 @@
...
+	case *PrivateKey[G1]:
+		var Q GG.G2
+		Q.Hash(msg, []byte(dstG2))
+		Q.ScalarMult(&k.key, &Q)
+		return Q.BytesCompressed()
+	case *PrivateKey[G2]:
+		var Q GG.G1
+		Q.Hash(msg, []byte(dstG1))
+		Q.ScalarMult(&k.key, &Q)
+		return Q.BytesCompressed()
```
Comment (bwesterb) [inline sign/bls/bls.go:276]: The compression matches the spec also in the corner cases?

File: sign/bls/bls.go [hunk 6]
```diff
@@ -0,0 +1,414 @@
...
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+// VerifyAggregate returns true if the aggregated signature is valid for
+// the list of messages and public keys provided. The slices must have
+// equal size and have at least one element.
+func VerifyAggregate[K KeyGroup](pubs []*PublicKey[K], msgs [][]byte, aggSig Signature) bool {
+	if len(pubs) != len(msgs) || len(pubs) == 0 || len(msgs) == 0 {
```
Comment (bwesterb) [inline sign/bls/bls.go:364]: You can leave out `lem(msgs) == 0`

File: sign/bls/bls.go [hunk 7]
```diff
@@ -0,0 +1,414 @@
...
+		aa, bb := new(G1), new(G1)
+		a, b = aa, bb
+		k := (interface{})(pub.key).(G2)
+		listG2[0], listG2[1] = &k.g, GG.G2Generator()
+		listG1[0], listG1[1] = &aa.g, &bb.g
+	default:
+		panic(ErrInvalid)
+	}
+
+	err := b.setBytes(sig)
```
Comment (bwesterb) [inline sign/bls/bls.go:311]: Does this do the subgroup check?

File: sign/bls/bls.go [hunk 8]
```diff
@@ -0,0 +1,414 @@
...
+
+// VerifyAggregate returns true if the aggregated signature is valid for
+// the list of messages and public keys provided. The slices must have
+// equal size and have at least one element.
+func VerifyAggregate[K KeyGroup](pubs []*PublicKey[K], msgs [][]byte, aggSig Signature) bool {
+	if len(pubs) != len(msgs) || len(pubs) == 0 || len(msgs) == 0 {
+		return false
+	}
+
+	n := len(pubs)
```
Comment (bwesterb) [inline sign/bls/bls.go:368]: I think you forgot the public key validation.

File: sign/bls/bls.go [hunk 9]
```diff
@@ -0,0 +1,414 @@
...
+
+	default:
+		panic(ErrInvalid)
+	}
+}
+
+// VerifyAggregate returns true if the aggregated signature is valid for
+// the list of messages and public keys provided. The slices must have
+// equal size and have at least one element.
+func VerifyAggregate[K KeyGroup](pubs []*PublicKey[K], msgs [][]byte, aggSig Signature) bool {
```
Comment (bwesterb) [inline sign/bls/bls.go:363]: This is `CoreAggregateVerify`, right?

File: sign/bls/bls.go [hunk 10]
```diff
@@ -0,0 +1,414 @@
...
+	}
+
+	n := len(pubs)
+	listG1 := make([]*GG.G1, n+1)
+	listG2 := make([]*GG.G2, n+1)
+	listExp := make([]int, n+1)
+
+	listG1[n] = GG.G1Generator()
+	listG2[n] = GG.G2Generator()
+	listExp[n] = -1
```
Comment (bwesterb) [inline sign/bls/bls.go:375]: `signs` is what `ProdPairFrac` uses.

**Review by bwesterb — CHANGES_REQUESTED** [review]

File: sign/bls/bls.go [hunk 11]
```diff
@@ -0,0 +1,414 @@
...
+//
+// # Serialization
+//
+// The serialization of elements in G1 and G2 follows the recommendation
+// given in [2], in order to be compatible with other implementations of
+// BLS12-381 curve.
+//
+// # References
+//
+// [1] https://datatracker.ietf.org/doc/draft-irtf-cfrg-bls-signature/
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go]: done

File: sign/bls/bls.go [hunk 12]
```diff
@@ -0,0 +1,414 @@
...
+	return k.pub
+}
+
+func (k *PrivateKey[K]) Equal(x crypto.PrivateKey) bool {
+	xx, ok := x.(*PrivateKey[K])
+	if !ok {
+		return false
+	}
+
+	switch (interface{})(k).(type) {
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go]: nice, much simpler

File: sign/bls/bls.go [hunk 13]
```diff
@@ -0,0 +1,414 @@
...
+		rd := hkdf.New(sha256.New, ikmZero, salt, keyInfoTwo)
+		n, err := io.ReadFull(rd, OKM)
+		if n != len(OKM) || err != nil {
+			return nil, err
+		}
+
+		ss.SetBytes(OKM)
+
+		if ss.IsZero() == 1 {
+			digest := sha256.Sum256(salt)
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go:249]: no, this branch happens when HKDF outputs all-zeros, or a multiple of a prime.

File: sign/bls/bls.go [hunk 14]
```diff
@@ -0,0 +1,414 @@
...
+			return &PrivateKey[K]{key: ss, pub: nil}, nil
+		}
+	}
+
+	return nil, ErrKeyGen
+}
+
+// Sign computes a signature of a message using a key (defined in
+// G1 or G1).
+func Sign[K KeyGroup](k *PrivateKey[K], msg []byte) Signature {
```
[the remark this answers is not in the document]
Comment (armfazh) [reply sign/bls/bls.go:261]: yes

File: sign/bls/bls.go (hunk 5 above)
Comment (armfazh) [reply sign/bls/bls.go:276]: all the points have fixed-size length. or do you mean something different?

File: sign/bls/bls.go (hunk 6 above)
Comment (armfazh) [reply sign/bls/bls.go]: true

File: sign/bls/bls.go (hunk 7 above)
Comment (armfazh) [reply sign/bls/bls.go:311]: yes

File: sign/bls/bls.go (hunk 8 above)
Comment (armfazh) [reply sign/bls/bls.go:374]: added now

File: sign/bls/bls.go (hunk 9 above)
Comment (armfazh) [reply sign/bls/bls.go:363]: yes, for the basic mode the core<name> functions are equal to <name>.

File: sign/bls/bls.go (hunk 10 above)
Comment (armfazh) [reply sign/bls/bls.go]: fixed

File: sign/bls/bls.go (hunk 5 above)
Comment (bwesterb) [reply sign/bls/bls.go:276]: Case of point at infinity.

**Review by bwesterb — APPROVED** [review]
Please address the test for point for infinity. Otherwise it's ok.

File: sign/bls/bls.go (hunk 5 above)
Comment (armfazh) [reply sign/bls/bls.go:276]: Tests for serialization here depend only on the unique ciphersuite: BLS12-381 curve
The encoding format is tested in the ecc/bls12381 package, see: https://github.com/cloudflare/circl/blob/main/ecc/bls12381/encoding_test.go#L73

**Comment by armfazh** [comment]
> Please address the test for point for infinity. Otherwise it's ok.

Tests for the point at infinity are checked in the [ecc/bls12831 package](https://github.com/cloudflare/circl/blob/main/ecc/bls12381/encoding_test.go#L73)
